From: Kefu Chai <kchai@redhat.com>
Date: Fri, 11 Nov 2016 21:37:50 +0800
Subject: ReplicatedPG::failed_push: release read lock on failure

and requeue the blocked ops.

Fixes: http://tracker.ceph.com/issues/17857
Signed-off-by: Kefu Chai <kchai@redhat.com>
(cherry picked from commit b3224a18f6acc7ed54c2162b140a33b6146a16be)
(cherry picked from commit 1926adc826de9ddaaa04138ce2a0b276eeda7ad1)

Resolves: rhbz#1414613
---
 src/osd/ReplicatedPG.cc | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/osd/ReplicatedPG.cc b/src/osd/ReplicatedPG.cc
index c43565b..89d243d 100644
--- a/src/osd/ReplicatedPG.cc
+++ b/src/osd/ReplicatedPG.cc
@@ -9590,7 +9590,14 @@ void ReplicatedPG::recover_got(hobject_t oid, eversion_t v)
 
 void ReplicatedPG::failed_push(const list<pg_shard_t> &from, const hobject_t &soid)
 {
+  dout(20) << __func__ << ": " << soid << dendl;
   assert(recovering.count(soid));
+  auto obc = recovering[soid];
+  if (obc) {
+    list<OpRequestRef> blocked_ops;
+    obc->drop_recovery_read(&blocked_ops);
+    requeue_ops(blocked_ops);
+  }
   recovering.erase(soid);
   for (auto&& i : from)
     missing_loc.remove_location(soid, i);
